Sources/Hummingbird/Router/Trie/Trie+resolve.swift

```diff
@@ -18,130 +18,197 @@ extension RouterTrie {
     /// Resolve a path to a `Value` if available
     @inlinable
     public func resolve(_ path: String) -> (value: Value, parameters: Parameters)? {
-        let pathComponents = path.split(separator: "/", omittingEmptySubsequences: true)
-        var pathComponentsIterator = pathComponents.makeIterator()
-        var parameters = Parameters()
+        var context = ResolveContext(path: path, trie: trie, values: values)
+        return context.resolve()
+    }
+
+    @usableFromInline
+    struct ResolveContext {
+        @usableFromInline let path: String
+        @usableFromInline let pathComponents: [Substring]
+        @usableFromInline let trie: Trie
+        @usableFromInline let values: [Value?]
+        @usableFromInline var parameters = Parameters()
+
+        @usableFromInline init(path: String, trie: Trie, values: [Value?]) {
+            self.path = path
+            self.trie = trie
+            self.pathComponents = path.split(separator: "/", omittingEmptySubsequences: true)
+            self.values = values
+        }
+
+        @usableFromInline func nextPathComponent(advancingIndex index: inout Int) -> Substring? {
+            if index >= pathComponents.count {
+                return nil
+            }
+
+            let component = pathComponents[index]
+            index += 1
+            return component
+        }
 
-        var node = trie.nodes[0]
-        var nodeIndex = 1
+        @inlinable
+        mutating func resolve() -> (value: Value, parameters: Parameters)? {
+            guard let component = pathComponents.first else {
+                guard let value = values[trie.nodes[0].valueIndex] else {
+                    return nil
+                }
+
+                return (value: value, parameters: parameters)
+            }
+
+            var nextNodeIndex = 1
+            guard let node = descend(
+                component: component,
+                nextPathComponentIndex: 1,
+                nextNodeIndex: &nextNodeIndex
+            ) else {
+                return nil
+            }
+
+            if let value = values[node.valueIndex] {
+                return (value: value, parameters: parameters)
+            } else {
+                return nil
+            }
+        }
+
+        @inlinable
+        mutating func descend(
+            component: Substring,
+            nextPathComponentIndex: Int,
+            nextNodeIndex: inout Int
```

Comment:
`nextNodeIndex` seems a little bit of a misnomer here. Given you are matching `component` shouldn't it just be `nodeIndex`.

I understand why `nextPathComponentIndex` had the next prefix as in it is the index of the next path component after `component`, but not sure it makes sense for `nextNodeIndex`.

Reply:
Depends a bit on where in the code's flow you're currently at whether or not it's the current component or the next one.